plugins/config.py

```diff
@@ -0,0 +1,153 @@
+from bob.errors import ParseError
+from bob.input import PluginState, PluginProperty
+import schema
+
+class TwoInts:
+    def validate(self, data):
+        data = schema.Schema([int]).validate(data)
+        if len(data) != 2:
+            raise schema.SchemaError("range property must have two integers")
+        return data
+
+COMMON_SCHEMA = schema.Schema({
+    schema.Optional('type') : schema.Or("choice", "bool", "str", "hex", "int"),
+    schema.Optional('required') : bool,
+    str : object,
+})
+
+CHOICE_SCHEMA = schema.Schema({
+    'type' : 'choice',
+    schema.Optional('required') : bool,
+    schema.Optional('help') : str,
+    'choice' : schema.Schema({
+        str : schema.Or(None, { schema.Optional('help') : str })
+    })
+})
+
+BOOL_SCHEMA = schema.Schema({
+    'type' : 'bool',
+    schema.Optional('required') : bool,
+    schema.Optional('help') : str,
+})
+
+STR_SCHEMA = schema.Schema({
+    schema.Optional('type') : 'str',
+    schema.Optional('required') : bool,
+    schema.Optional('help') : str,
+})
+
+HEX_SCHEMA = schema.Schema({
+    'type' : 'hex',
+    schema.Optional('required') : bool,
+    schema.Optional('help') : str,
+    schema.Optional('range') : TwoInts(),
+})
+
+INT_SCHEMA = schema.Schema({
+    'type' : 'int',
+    schema.Optional('required') : bool,
+    schema.Optional('help') : str,
+    schema.Optional('range') : TwoInts(),
+})
+
+ALL_SCHEMA = {
+    'choice' : CHOICE_SCHEMA,
+    'bool' : BOOL_SCHEMA,
+    'str' : STR_SCHEMA,
+    'hex' : HEX_SCHEMA,
+    'int' : INT_SCHEMA,
+}
+
+class VarValidator:
+    def validate(self, data):
+        data = COMMON_SCHEMA.validate(data)
+        return ALL_SCHEMA[data.get('type', "str")].validate(data)
+
+SCHEMA = schema.Schema({ str : VarValidator() })
+
+class ConfigProperty(PluginProperty):
+    @staticmethod
+    def validate(data):
+        return SCHEMA.validate(data)
+
+    def inherit(self, cls):
+        if self.present:
+            if cls.present:
+                # Take only new values from class
+                self.value.update({ name : descr for name, descr in cls.value.items()
+                                    if name not in self.value })
+        else:
+            self.present = cls.present
+            self.value = cls.value
+
+
+def handleChoice(var, val):
+    if val not in var["choice"]:
+        return f"Invalid 'choice': key '{val}' not defined"
+
+def handleBool(var, val):
+    if val not in ("0", "1"):
+        return f"'{val}' is not a boolean"
+
+def handleStr(var, val):
+    pass
+
+def handleHex(var, val):
```

Comment:
Can't we handle that in the int function below. I am not a fan of having dedicated methods for converting those.

Reply:
Hmm, the error strings are different too. Not sure if we gain too much to deduplicate the remaining code.